### PR TITLE
[QuestionService] Add room feature

### PR DESCRIPTION
**Pull request by eugenetyc**
Adds the `GET /question/room` and `POST /question/room` endpoints.

- `POST /question/room` creates the roomId to questionId mapping in Question Service. Mapping between the 2 users will be refreshed with a potentially new questionId value (due to randomized tie-break nature) each time this is called.
- `GET /question/room` takes in the roomId supplied and provides a questionId to whoever is calling. We expect only 2 users to know of this roomId, which is a hash generated in the POST endpoint.

- Removed `POST /question/match` as it is now redundant. Note that certain question add/delete methods are still present but in comments, in case we want to allow users to add questions. Should remove if we are not planning to.

- Added tests and updated dummy question data with real algorithm questions from leetcode. However, these do not cover all the topics and difficulties combinations; instead they allow for us to be sure that our tests can cover every functionality.

### CODE REVIEW

File: QuestionService/application/service/room-service.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Application Layer (Service) defines the HTTP Route Handler Functions
+ * All Functions defined here will have `req` and `res`
+ * These contain the function definition of the HTTP Routes
+ */
+
+const CreateRoom = async (req, res) => {
+  try {
+    const { topics, difficulties, username1, username2 } = req.body
+    const respOrm = await ormRoom.CreateRoom(topics, difficulties, username1, username2)
+    return await wrapResult(res, 'Cannot Create Room', 'Room Created', respOrm)
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:14]: done

[the remark this answers is not in the document]
Comment (DivineDX) [reply QuestionService/application/service/room-service.js:14]: > done

Hmm has it been pushed?

File: QuestionService/util/response.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * HTTP Response Template
+ * @param {String} status
+ * @param {String} message
+ * @param {Array} data
+ */
+exports.Response = (status, message, data = []) => ({
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/util/response.js:7]: done

File: QuestionService/application/service/room-service.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  } catch (err) {
+    console.log('err: ', err)
+    const resp = await Response('Failure', 'DB failed', [])
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:17]: done

File: QuestionService/application/service/room-service.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+const FindRoomById = async (req, res) => {
+  try {
+    const { roomId } = req.body
+    const respOrm = await ormRoom.FindRoom(roomId)
+    return await wrapResult(res, 'Cannot Find Room', 'Found Room', respOrm)
+  } catch (err) {
+    console.log('err: ', err)
+    const resp = await Response('Failure', 'DB failed', [])
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:29]: done

File: QuestionService/application/orm/room-orm.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+// for generating room id
+const saltRounds = 3
+const saltRounds2 = 2
+const salt = bcrypt.genSaltSync(saltRounds)
+const salt2 = bcrypt.genSaltSync(saltRounds2)
+
+exports.CreateRoom = async (topics, difficulties, username1, username2) => {
+  try {
+    if (!topics || !difficulties || !username1 || !username2) {
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/room-orm.js:13]: done

File: QuestionService/infrastructure/persistence/room-repository.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+// Room
+
+const roomsDb = db.rooms
+
+const findRoom = async (condition) => roomsDb.findOne(condition)
+
+const deleteRoom = async (del) => roomsDb.deleteMany(del)
+
+const createRoom = async (del, params) =>
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/infrastructure/persistence/room-repository.js:13]: done

File: QuestionService/application/orm/room-orm.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+      throw new Error('Request has missing required attribute(s)')
+    }
+
+    const beforeHash = `${salt}${username1}${salt2}${username2}`
+    const hash = String(bcrypt.hashSync(beforeHash, salt))
+
+    const matchedQuestion = await FindMatchedQuestionUtility(topics, difficulties)
+
+    const deleteCondition = { roomId: hash }
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/room-orm.js:22]: yes

File: QuestionService/application/orm/room-orm.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+const roomsRepo = require('../../infrastructure/persistence/room-repository')
+const { FindMatchedQuestionUtility } = require('./question-orm')
+
+// for generating room id
+const saltRounds = 3
+const saltRounds2 = 2
+const salt = bcrypt.genSaltSync(saltRounds)
+const salt2 = bcrypt.genSaltSync(saltRounds2)
+
+exports.CreateRoom = async (topics, difficulties, username1, username2) => {
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/room-orm.js:11]: done!

[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/room-orm.js:11]: good point on 

> Remember that ORM is for modelling and interacting with the DB, we should not care about business logic here.

File: QuestionService/application/orm/question-orm.js [hunk 9]
```diff
@@ -1,5 +1,22 @@
...
+    throw new Error('No such question id exists')
+  }
+
+  const randomIndex = getRandomNumberBetweenInclMinExclMax(0, results.length)
+  const randomResult = results[randomIndex]
+
+  return randomResult
+}
+
+exports.FindMatchedQuestionUtility = FindMatchedQuestionUtility
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/question-orm.js:18]: done

File: QuestionService/application/service/room-service.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    const respOrm = await ormRoom.CreateRoom(topics, difficulties, username1, username2)
+    return await wrapResult(res, 'Cannot Create Room', 'Room Created', respOrm)
+  } catch (err) {
+    console.log('err: ', err)
+    const resp = await Response('Failure', 'DB failed', [])
+    return res.status(500).send(resp)
+  }
+}
+
+const FindRoomById = async (req, res) => {
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:22]: Can, I was thinking of this but decided to stick to questionId as per agreed

[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:22]: #TODO

File: QuestionService/application/service/room-service.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:14]: later

File: QuestionService/application/orm/question-orm.js [hunk 11]
```diff
@@ -1,5 +1,22 @@
...
+
+// to filter questions for matching
+const FindMatchedQuestionUtility = async (topics, difficulties) => {
+  const filter = { topics, difficulties }
+  const results = await questionsRepo.findMatch(filter)
+  if (!results || results.length === 0) {
+    throw new Error('No such question id exists')
+  }
+
+  const randomIndex = getRandomNumberBetweenInclMinExclMax(0, results.length)
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/question-orm.js:12]: done

File: QuestionService/application/service/room-service.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/service/room-service.js:22]: Currently changed to:

`    const replacement = {
      roomId: hash,
      questionId: randomMatchedQuestion.id,
      question: randomMatchedQuestion,
    }`

File: QuestionService/test/endpoint.test.js [hunk 12]
```diff
@@ -41,4 +79,276 @@ describe('Endpoint Testing', () => {
...
+    const matchRequestData = {
+      username1: 'firebreathingeugene',
+      username2: 'waterbreathingeugene',
+      topics: ['Arrays'],
+      difficulties: ['Easy', 'Medium', 'Difficult'],
+    }
+    const getRoomIdResult = await chai
+      .request(app)
+      .post('/question/room')
+      .set('content-type', 'application/x-www-form-urlencoded')
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/test/endpoint.test.js:211]: done

File: QuestionService/application/orm/room-orm.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  try {
+    if (!topics || !difficulties || !username1 || !username2) {
+      throw new Error('Request has missing required attribute(s)')
+    }
+
+    const beforeHash = `${salt}${username1}${salt2}${username2}`
+    const hash = String(bcrypt.hashSync(beforeHash, salt))
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/application/orm/room-orm.js:18]: if hash is too easy, others can access and spoil the user experience though; but then again do we want users to type this? I remember we wanted to possibly have this button at matching page for users to reconnect if they need right? Maybe that can automate this?

Also, I realised sometimes for same pair of users, the hash is different using bcrypt. Looking into an alternative;

File: QuestionService/test/endpoint.test.js [hunk 14]
```diff
@@ -41,4 +79,276 @@ describe('Endpoint Testing', () => {
...
+      .send(emptyMatchRequestTopicsData)
+
+    VerifyFailure(emptyMatchRequestTopicsDataResult, 404)
+    const emptyTopicsBody = emptyMatchRequestTopicsDataResult.body
+    emptyTopicsBody.status.should.eql('fail')
+    emptyTopicsBody.message.should.eql('Cannot Create Room')
+    emptyTopicsBody.data.should.be.a('object')
+    emptyTopicsBody.data.should.have.property('err')
```
[the remark this answers is not in the document]
Comment (eugenetyc) [reply QuestionService/test/endpoint.test.js:278]: done

File: QuestionService/test/utils.test.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
+const chai = require('chai')
+const { STATUS_SUCCESS, STATUS_FAIL } = require('../util/enums')
+const app = require('../server')
+
+const dummyQuestionData = require('../data/dummy-questions.json')
+const { questionsDb } = require('../infrastructure/persistence/question-repository')
+
+chai.should()
```
Comment (DivineDX) [inline QuestionService/test/utils.test.js:8]: nice file! I should probably do this too :P

**Review by DivineDX — APPROVED** [review]

File: QuestionService/application/service/room-service.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (DivineDX) [reply QuestionService/application/service/room-service.js:39]: Yup as discussed online, can stick to just returning questionId

**Pull request merged**